### ISSUE TITLE
fcu_io crashes

### ISSUE THREAD

**Issue by superjax** [BYU-MAGICC/fcu_io2#15]
I have now had 3 times (out of about 15 times) where fcu_io stops responding.  It's not obvious to me why, or what causes it.  It appears that the node completely freezes (Ctrl+C doesn't do anything).  The naze is still responding, with RC control, so I think it's safe to flight test, but I think we need to make sure that doesn't happen.

I'll keep digging, and see if I can figure out what is causing it.

Any ideas @dpkoch?

**Comment by dpkoch** [comment]
I haven't had anything like this happen to me. I've had the node just running and streaming IMU for up to an hour before with no issue. So I need to figure out how to replicate this before I can debug.

One thought; were you sending offboard control commands when it froze? That's one thing I haven't done much of so that could potentially be the difference.

**Comment by superjax** [comment]
Yeah, I have had it happen while testing HIL-style.  (extended commands coming from Gazebo going to the plane, but streaming senors at full rate from the plane to a bogus topic).  When it freezes, all MAVlink comes to a halt. 

I've been trying to replicate it, and I haven't had a lot of consistent success.  I've been able to do it again but when I can come up with a recipe for breaking we'll have a little easier time pinning down the problem.

**Comment by superjax** [comment]
Okay, so I think I might know what is going on.   I believe that perhaps if the mavlink message callbacks take too long, then a buffer overflows somewhere.

I noticed it when I was trying to implement the accelerometer calibration and I was doing a least-squares like an idiot on 100,000 measurements.  It took the computer about 3 seconds to do all the axes, and then fcu_io would promptly freeze.   I have since throttle the measurements so we aren't doing so many for the least-squares, but I just tossed this code into the `handle_small_imu_msg` function and can replicate the bug.

``` c
void fcuIO::handle_small_imu_msg(const mavlink_message_t &msg)
{
  static int bug_counter = 0;

  if (bug_counter == 1000)
  {
    ROS_ERROR("BUG TESTING, TAKING A BREAK FOR 2 SECONDS");
    sleep(2);
  }
  else if(bug_counter > 1000)
  {
    ROS_ERROR("BUG DIDN'T FREEZE");
  }
  bug_counter++;
  ...
```

Interestingly, `BUG DIDN'T FREEZE` runs about 5 times, but then the program halts and won't respond.  Any ideas?
